session/general_log.go

```diff
@@ -0,0 +1,196 @@
+}
+
+// StopGeneralLog stops the general log worker goroutine
+func StopGeneralLog() {
+	globalGeneralLogger.stopLogWorker()
+}
+
+// generalLogEntry represents the fields in a general query log
+type generalLogEntry struct {
+	ConnID            uint64
+	User              string
+	SchemaMetaVersion int64
+	TxnStartTS        uint64
+	TxnForUpdateTS    uint64
+	IsReadConsistency bool
+	CurrentDB         string
+	TxnMode           string
+	Query             generalLogEntryQuery
+}
+type generalLogEntryQuery struct {
+	isPrepared      bool
+	originalText    string
+	stmtNode        ast.StmtNode
+	stmtCtx         *stmtctx.StatementContext
+	preparedParams  variable.PreparedParams
+	enableRedactLog bool
+}
+
+// getGeneralLogEntry get a general log entry from the object pool
+func getGeneralLogEntry() *generalLogEntry {
+	return glEntryPool.Get().(*generalLogEntry)
+}
+
+// generalLogger receives general log entry from channel, and log or drop them as needed
+type generalLogger struct {
+	logger       *zap.Logger
+	logEntryChan chan *generalLogEntry
+	quit         chan struct{}
+}
+
+func newGeneralLogger(logger *zap.Logger) *generalLogger {
+	gl := &generalLogger{
+		logger:       logger,
+		logEntryChan: make(chan *generalLogEntry, 10000),
+		quit:         make(chan struct{}),
+	}
+	go gl.startLogWorker()
+	return gl
+}
+
+func (gl *generalLogger) logEntry(e *generalLogEntry) {
+	buf := queryBuilderPool.Get().(*strings.Builder)
+	q := e.Query
+	if q.isPrepared {
+		buf.WriteString(q.originalText)
+	} else {
+		// The following logic should be kept in sync with ExecStmt.GetTextToLog
+		if q.enableRedactLog {
+			sql, _ := q.stmtCtx.SQLDigest()
+			buf.WriteString(sql)
+		} else if sensitiveStmt, ok := q.stmtNode.(ast.SensitiveStmtNode); ok {
+			buf.WriteString(sensitiveStmt.SecureText())
+		} else {
+			buf.WriteString(q.stmtCtx.OriginalSQL)
+			buf.WriteString(q.preparedParams.String())
+		}
+	}
+	bufString := buf.String()
+	queryMutable := *(*[]byte)(unsafe.Pointer(&bufString))
+	for i, b := range queryMutable {
+		if b == '\r' || b == '\n' || b == '\t' {
+			queryMutable[i] = ' '
+		}
+	}
+	gl.logger.Info("GENERAL_LOG",
+		zap.Uint64("conn", e.ConnID),
+		zap.String("user", e.User),
+		zap.Int64("schemaVersion", e.SchemaMetaVersion),
+		zap.Uint64("txnStartTS", e.TxnStartTS),
+		zap.Uint64("forUpdateTS", e.TxnForUpdateTS),
+		zap.Bool("isReadConsistency", e.IsReadConsistency),
+		zap.String("current_db", e.CurrentDB),
+		zap.String("txn_mode", e.TxnMode),
+		zap.String("sql", buf.String()),
+	)
+	buf.Reset()
+	queryBuilderPool.Put(buf)
+	glEntryPool.Put(e)
+}
+
+// startLogWorker starts a log flushing worker that flushes log periodically or when batch is full
+func (gl *generalLogger) startLogWorker() {
+	for {
+		select {
+		case e := <-gl.logEntryChan:
+			gl.logEntry(e)
+		case <-gl.quit:
+			for e := range gl.logEntryChan {
+				gl.logEntry(e)
+			}
+			return
+		}
+	}
+}
+
+func (gl *generalLogger) stopLogWorker() {
+	close(gl.quit)
```

Comment:
I guess this `close()` might not be a blocking call, so that it may be possible that the server shutdown before everything is actually consumed. Could you revisit this part?

Reply:
Fixed with a `sync.WaitGroup`.